sdk/core/azure-core-amqp/inc/azure/core/amqp/internal/common/runtime_context.hpp

```diff
@@ -134,5 +134,28 @@ namespace Azure { namespace Core { namespace Amqp { namespace Common { namespace
     builder.reset(raw);
   }
 
+  /**
+   * Invoke a simple Rust AMQP API, checking for error.
+   *
+   * @param api - Flat C API to invoke. The first parameter MUST be a RustCallContext object, the
+   * second parameter must be a Rust client object.
+   * @param amqpObject - Unique Pointer to a Rust AMQP object.
+   * @param args - remaining arguments to the flat C API.
+   *
+   * This function will check the return from the API, and if it is not 0, will throw an exception
+   * with error information from the callContext.
+   *
+   */
+  template <typename Api, typename T, typename... Args>
+  void InvokeAmqpApi(Api& api, T& amqpObject, Args&&... args)
```

Comment:
I think if you wanted, you could force types a bit more strictly, for example for `amqpObject` you could put its type as `unique_ptr<T>`
```suggestion
  void InvokeAmqpApi(Api& api, std::unique_ptr<T> const& amqpObject, Args&&... args)
```
But if we want it to work with `shared_ptr`s as well, then it will end up being more cumbersome.
Or the opposite - if we want it to work with raw pointers, you could use `*amqpObject` instead of `amqpObject.get()`, and it will work with both smart and raw pointers.

This is a non-urgent thing for sure, it lives in `_detail`, and can be updated at any time later, if needed, and please only do if you agree, I am just proposing what I seem to see but you have better knowledge of the context, feel free to not take this suggestion.

Reply:
I specifically *don't* want to work with raw pointers - it's part of the point of this function is that it allows the caller to avoid all the `.get()` calls. 

And because of the custom deleter which each of these types uses, `std::unique_ptr<T>` cannot be used as written :(.

However with a bit of tweaking, I think I have something that works.